site/content/blog/multi-env-deployment-ytt-kapp.md

```diff
@@ -0,0 +1,331 @@
+
+```yaml
+#! schema.yaml (#! is used for comment in ytt)
+#@data/values-schema
+---
+replicaCount: 1
+
+image:
+  name: nginx
+  tag: "1.14.2"
+
+nameOverride: ""
+fullnameOverride: ""
+
+appMode: ""
+certificatePath: ""
+databaseUser: ""
+databasePassword: ""
+```
+
+While normally schema defines the default configuration, it makes sense to create different values files for all different environments. Let's suppose we have two environments `staging` and `prod`.
+
+In `values.yaml` we will be just putting the values that we want to override from schema
+ 
+```yaml
+#! values-staging.yaml
+#@data/values
+---
+image:
+  tag: "latest"
+
+nameOverride: "staging"
+
+appMode: staging
+certificatePath: /etc/ssl/staging
+databaseUser: staging-user
+databasePassword: staging-password
+```
+
+In production maybe we want to have more replicas so we can override that in `values-prod.yaml`
+
+```yaml
+#! values-prod.yaml
+#@data/values
+---
+replicaCount: 3
+
+nameOverride: "prod"
+
+appMode: prod
+certificatePath: /etc/ssl/prod
+databaseUser: prod-user
+databasePassword: prod-password
+```
+
+## Using Data Values in manifest - 
+Now we will see how we can use the data values that has been declared via data value schema. Here I have put all the resources in single yaml file `app.yaml` just for the demonstration purpose.
+
+```yaml
+#! app.yaml
+#@ load("@ytt:data", "data")
+---
+#@ def fullname():
+#@ return data.values.fullnameOverride if data.values.fullnameOverride else "sample-app-" + data.values.nameOverride
+#@ end
+
+#@ def labels():
+app.kubernetes.io/name: #@ fullname()
+#@ end
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ fullname()
+  labels: #@ labels()
+spec:
+  selector:
+    app: #@ fullname()
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 8080
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: #@ fullname()
+  labels: #@ labels()
+spec:
+  replicas: #@ data.values.replicaCount
+  selector:
+    matchLabels: #@ labels()
+  template:
+    metadata:
+      labels: #@ labels()
+    spec:
+      containers:
+        - name: sample-app
+          image: #@ data.values.image.name + ":" + data.values.image.tag
+          imagePullPolicy: IfNotPresent
+          ports:
+          - containerPort: 8080
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: application-settings
+data:
+  app_mode: #@ data.values.appMode
+  certificates: #@ data.values.certificatePath
+  db_user: #@ data.values.databaseUser
+  db_password: #@ data.values.databasePassword
+```
+
+As you can see above file contains lots of ytt annotations (i.e. lines that contain #@), it’s a ytt template. With the help of this annotations we are using the values defined in schema file. 
```

Comment:
I can't put my finger on it why, but this paragraph delights me. It's clear, on point, and it an invitation to the reader to really dig into the example above rather than explain it to them. 👍🏻